src/UserSpaceInstrumentation/AccessTraceesMemory.cpp

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "AccessTraceesMemory.h"
+
+#include <absl/base/casts.h>
+#include <absl/strings/numbers.h>
+#include <absl/strings/str_format.h>
+#include <absl/strings/str_split.h>
+#include <sys/ptrace.h>
+
+#include <cerrno>
+#include <cstring>
+#include <string>
+
+#include "OrbitBase/Logging.h"
+#include "OrbitBase/ReadFileToString.h"
+#include "OrbitBase/SafeStrerror.h"
+
+namespace orbit_user_space_instrumentation {
+
+using orbit_base::ReadFileToString;
+
+[[nodiscard]] ErrorMessageOr<std::vector<uint8_t>> ReadTraceesMemory(pid_t pid,
+                                                                     uint64_t address_start,
+                                                                     uint64_t length) {
+  // Round up length to next multiple of eight.
+  length = ((length + 7) / 8) * 8;
+  std::vector<uint8_t> bytes(length);
+  for (size_t i = 0; i < length / 8; i++) {
+    const uint64_t data = ptrace(PTRACE_PEEKDATA, pid, address_start + i * 8, NULL);
+    if (errno) {
+      return ErrorMessage(
+          absl::StrFormat("Failed to PTRACE_PEEKDATA for pid: %d with errno %d: \"%s\"", pid, errno,
+                          SafeStrerror(errno)));
+    }
+    std::memcpy(bytes.data() + (i * sizeof(uint64_t)), &data, 8);
+  }
+  return bytes;
+}
+
+[[nodiscard]] ErrorMessageOr<void> WriteTraceesMemory(pid_t pid, uint64_t address_start,
+                                                      const std::vector<uint8_t>& bytes) {
+  size_t pos = 0;
+  do {
+    // Pack 8 bytes for writing into `data`.
+    uint64_t data = 0;
+    std::memcpy(&data, bytes.data() + pos, sizeof(uint64_t));
+    if (ptrace(PTRACE_POKEDATA, pid, address_start + pos, data) == -1) {
+      ErrorMessage(absl::StrFormat("Failed to PTRACE_POKEDATA for pid: %d.", pid));
+    }
+    pos += sizeof(uint64_t);
+  } while (pos < bytes.size());
+  return outcome::success();
+}
+
+[[nodiscard]] ErrorMessageOr<void> GetFirstExecutableMemoryRegion(pid_t pid, uint64_t* addr_start,
```

Comment:
I think Dimitry mentioned in the previous change, but have you considered returning a `pair` or a dedicated `struct` instead of using two out parameters?

Reply:
Funny! The style guide evolved a lot from what I remembered. This was the recommended way of doing things up until 2017. Let's go with a type alias for a pair.